Bring BioCompute exports under invocation export framework

jmchilton opened the pull request:
- Builds on RO Crate Export
- Task-based and a similar API should make it easier to integrate into #14606 
- Adds validation previously missing using data model generation approach from DRS work ( #13949 )
- Much more typing throughout.
- Formal annotation of request parameters and such in OpenAPI.

WIP: needs some tests - but everything is strongly typed so the pieces should largely fit together very well.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bgruening commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2200,6 +2227,224 @@ def _init_crate(self):
+
+class BcoModelExportStore(WorkflowInvocationOnlyExportStore):
+    def __init__(self, uri, export_options: BcoExportOptions, **kwds):
+        temp_output_dir = tempfile.mkdtemp()
```
where is this tempdir created? I guess it will not be large and does not matter much, but should it be in the some Galaxy temp dir?

mvdbeek replied:
It will be part of new_file_path, we're patching the basedir when setting up `GalaxyAppConfiguration`

bgruening reviewed (commented):
Green :)

jmchilton committed:
Implement Workflow Invocation -> BCO export.

davelopez committed:
Add basic BCO export integration tests

davelopez committed:
Fix some pydantic validation issues

davelopez committed:
Exclude null values when serializing BCO

+ fix typo

davelopez committed:
Fix some BCO fields

- Pass the workflow url for the execution_domain instead of a fixed galaxy url
- Fix step annotations in PipelineStep description
- Use API url for downloading collections in InputSubdomainItem

nsoranzo commented on lib/galaxy/dependencies/pinned-requirements.txt:
```diff
@@ -55,6 +55,7 @@ docopt==0.6.2 ; python_version >= "3.7" and python_version < "3.11"
 docutils==0.16 ; python_version >= "3.7" and python_version < "3.11"
 ecdsa==0.18.0 ; python_version >= "3.7" and python_version < "3.11"
 edam-ontology==1.25.2 ; python_version >= "3.7" and python_version < "3.11"
+email-validator==1.3.0 ; python_version >= "3.7" and python_version < "3.11"
```
Seems unused? Also in other 2 places below.

davelopez replied:
Turns out it was used by pydantic :)

https://github.com/galaxyproject/galaxy/actions/runs/3217726750

nsoranzo commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1000,278 +997,6 @@ def show_invocation_report_pdf(self, trans: GalaxyWebTransaction, invocation_id,
-        return bco_dict
-
-    @expose_api
-    def export_invocation_bco(self, trans: GalaxyWebTransaction, invocation_id, **kwd):
```
Is it necessary to remove the API endpoints? Are we sure they are not used?
They are also exposed via BioBlend.

davelopez replied:
I think that was the whole point of the refactoring, unify the invocation export for all formats, but I see your point.
We can mark the old endpoints as deprecated for a while and point to the new endpoints in the docstring.

I'll see if I can reuse as much as possible from the new export system in the old endpoints to keep them minimal.

jmchilton replied:
>

        The BioCompute Object endpoints are in beta - important details such
        as how inputs and outputs are represented, how the workflow is encoded,
        and how author and version information is encoded, and how URLs are
        generated will very likely change in important ways over time.

We marked this as beta very explicitly and very intentionally. I would prefer to not maintain the endpoints or put @davelopez time on it but I'm not going to prevent anyone from doing that.

davelopez replied:
Upss, just pushed the temporal endpoints back :stuck_out_tongue_closed_eyes: in 8b50725.

It's kind of a hacky solution, so I'm happy to drop them or keep them for a while, @nsoranzo your thoughts?

nsoranzo commented on lib/galaxy/dependencies/pinned-requirements.txt:
```diff
@@ -55,6 +55,7 @@ docopt==0.6.2 ; python_version >= "3.7" and python_version < "3.11"
 docutils==0.16 ; python_version >= "3.7" and python_version < "3.11"
```
This is also required by email-validator and needs to be added above:
```
dnspython==2.2.1 ; python_version >= "3.7" and python_version < "3.11"
```

nsoranzo commented on pyproject.toml:
```diff
@@ -49,6 +49,7 @@ cwltool = "==3.1.20211107152837"
 dictobj = "*"
 docutils = "!=0.17, !=0.17.1"
 edam-ontology = "*"
+email-validator = "*"
```
I think it'd be more future-proof to replace this:

```suggestion
```

by modifying the pydantic requirement below to:

```
pydantic = {version = "*", extras = ["email"]}
```

as suggested in the import error raised in the API tests: ImportError: email-validator is not installed, run `pip install pydantic[email]`

nsoranzo commented on packages/data/setup.cfg:
```diff
@@ -38,6 +38,7 @@ install_requires =
     bdbag
     bx-python
     contextvars; python_version >= "3.6" and python_version < "3.7"
+    email-validator
```
```suggestion
```

and replace `pydantic` below with `pydantic[email]` .

davelopez replied:
Thank you @nsoranzo! I hope I did the correct changes in [5d824db](https://github.com/galaxyproject/galaxy/pull/14620/commits/5d824db1bb100b99b8d9f34a410cf3e3d3d9919c). Please take a look :)

nsoranzo replied:
Thank you, LGTM, let's see what the tests say :D

davelopez committed:
Use pydantic[email] instead of email-validator

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

davelopez committed:
Restore bco endpoints as deprecated

Minimal implementation trying to reuse the export store system to keep the old endpoints until they are finally removed.

nsoranzo commented on lib/galaxy_test/api/test_workflows.py:
```diff
@@ -2208,15 +2205,6 @@ def test_workflow_invocation_report_custom(self):
             assert "## About This Report" in markdown_content
 
-    @skip_without_tool("cat1")
-    def test_export_invocation_bco(self):
```
Should we also (temporarily) restore this test?

davelopez replied:
Sure! I can do that

nsoranzo replied:
Thanks!

davelopez replied:
While at it, I decided to validate the whole BCO against their schema as part of the tests in [c776c81](https://github.com/galaxyproject/galaxy/pull/14620/commits/c776c8169a9488a891bc71f05a3fd7e45b90d9b5) and we can keep that for future tests.

HadleyKing replied:
@davelopez 
How did that work out? The validation test? I know there are a few things that can not be pulled from Galaxy invocation and so it might always fail but if it is an expected validation fail that can still be a great test. Right?

davelopez replied:
@HadleyKing, the schema validation using https://opensource.ieee.org/2791-object/ieee-2791-schema/raw/master/2791object.json seems to validate fine for the workflow invocations under test (at least with the changes included in this PR). John Chilton generated the pydantic models from the schema and thanks to that found a couple of errors in the format that should be fixed now here.

Maybe there is still missing information, but the format itself returned by the API complies with the schema now.

davelopez committed:
Restore temporarily bco API test

davelopez committed:
Validate BioCompute Objects against schema in tests
